File: MATLAB/ASDM.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Code for the Sigma Delta
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

function [vOut,vInt,vError]= ASDM(x, schThrPerc, K, feedB, schOutLevel, timeStep, samples )

%-- System Parameters
schmittTriggerThreshold     = schThrPerc*schOutLevel; %- Switching threshold as a percentage

%IntegratorK                 = 3e-12; %- Integrator Constant (Capacitor value)

%%- Variables
lt        = samples;
vError    = zeros(1,lt); % Error function (Input-Feedback)
vFeedback = zeros(1,lt); % Feedback is just the output of the Schmitt Trigger
vInt      = zeros(1,lt); % Output of the integrator
vOut      = zeros(1,lt); % Output of the Modulator (Schmitt Trigger)

%%- Initial Conditions
vFeedback(1) = feedB*schOutLevel;

initCond = 0;
vInt(1)  = initCond;
for i=2:lt
  vError(i)    = x(i)-vFeedback(i-1);
  vInt(i)      = vInt(i-1) + (vError(i)*timeStep/K);
  vOut(i)      = schmittTrigger(vInt(i), vOut(i-1), schThrPerc, schOutLevel);
  vFeedback(i) = feedB*vOut(i);
end  


end

File: StochasticPIM/ASDM.cpp
#include <stdio.h>
#include <vector>
#include "ASDM.h"
#include <iostream>
#include "SchmittTrigger.h"
using namespace std;


//vals[0] = vOut, vals[1] = vInt, vals[2] = vError

float** ASDM(float* x, float schThrPerc, float K, float feedB, float schOutLevel, float timeStep, int samples){
    float schmittTriggerThreshold = schThrPerc*schOutLevel;
    float* vError = new float[samples];
    float* vFeedback = new float[samples];
    float* vInt = new float[samples];
    float* vOut = new float[samples];
    vFeedback[0] = feedB*schOutLevel;
    vInt[0] = 0;
    vError[0] = 0;
    vFeedback[0] = feedB*schOutLevel;
    vOut[0] = 0;
    for(int i = 1; i < samples; i++){
        vError[i] = (x[i]-vFeedback[i-1]);
        vInt[i] = (vInt[i-1]+(vError[i]*timeStep/K));
        vOut[i] = (schmittTrigger(vInt[i], vOut[i-1], schThrPerc, schOutLevel));
        vFeedback[i] = feedB*vOut[i];
    }
    float** vOuts = new float*[3];
    vOuts[0]= vOut;
    vOuts[1] = vInt;
    vOuts[2] = vError;

    return vOuts;
}
